src/GoogleFit/FitOnFhir.GoogleFit/Services/GoogleFitTokensService.cs

```diff
@@ -0,0 +1,52 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using EnsureThat;
+using FitOnFhir.Common.Exceptions;
+using FitOnFhir.Common.Repositories;
+using FitOnFhir.Common.Services;
+using FitOnFhir.GoogleFit.Client.Responses;
+using Microsoft.Extensions.Logging;
+
+namespace FitOnFhir.GoogleFit.Services
+{
+    public class GoogleFitTokensService : TokensServiceBase<AuthTokensResponse>
+    {
+        private readonly IGoogleFitAuthService _googleFitAuthService;
+        private readonly ILogger<GoogleFitTokensService> _logger;
+
+        public GoogleFitTokensService(
+            IGoogleFitAuthService googleFitAuthService,
+            IUsersKeyVaultRepository usersKeyVaultRepository,
+            ILogger<GoogleFitTokensService> logger)
+        : base(usersKeyVaultRepository, logger)
+        {
+            _googleFitAuthService = EnsureArg.IsNotNull(googleFitAuthService, nameof(googleFitAuthService));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+        }
+
+        public override async Task<AuthTokensResponse> RefreshToken(string googleFitId, CancellationToken cancellationToken)
+        {
+            AuthTokensResponse tokensResponse = null;
+
+            try
+            {
+                var refreshToken = await RetrieveRefreshToken(googleFitId, cancellationToken);
+
+                _logger.LogInformation("Refreshing the RefreshToken");
+                tokensResponse = await _googleFitAuthService.RefreshTokensRequest(refreshToken, cancellationToken);
```

Comment:
I think this class can be simplified significantly. I makes more sense to me for derived classes to not be concerned with fetching the token from the store and saving it, that logic should exist in the base only. Also note, that the only exception I am catching is very specific to the refresh operation failing. I want to avoid catching failures here that don't have anything to do with the refresh failing. In the base class, we can have a less restrictive try catch, where we catch most exceptions (assuming they are transient failures) and not catch the token refresh exception.

In the base class:
```c#
protected abstract Task<TTokenResponse> RefreshToken(string refreshToken, CancellationToken cancellationToken);
```

Implementation:
```c#
protected override async Task<AuthTokensResponse> RefreshToken(string refreshToken, CancellationToken cancellationToken)
{
    try
    {
        return await _googleFitAuthService.RefreshTokensRequest(refreshToken, cancellationToken);
    }
    catch (Google.Apis.Auth.OAuth2.Responses.TokenResponseException ex)
    {
        throw new TokenRefreshException(ex.Message);
    }
}
```